### ISSUE TITLE
Question: Is it possible to have multiple targets in a validation set

### ISSUE THREAD

**Issue by simpsus** [microsoft/LightGBM#5860]
I have a validation dataset that contains multiple targets. I want to train a regression and have several validation metrics, one for each of the targets in the validation set. At the moment I create the `valid_sets` parameter like

`valid_sets = [lgb.Dataset(data=df.filter(like="feature"), label=df[t]) for t in targets]`

which works, but I create a duplicate of the features each time. As a consequence, my memory is too small for my data.

Is it possible to define a validation set with multiple targets and have a metric for each of those reusing the features?

Thank you very much!

**Comment by jameslamb** [comment]
Can you tell us a bit more about what you mean by "multiple targets"? Do you just mean something like this?

| sales | region | ... other features ... |
|:-----:|:------:|:-------------------:|
| 10.0  |  West  | ...                               |
| 11.5  |  West  | ...                               |
| 23.1  |  East  | ...                               |
| 22.0  |  East  | ...                               |
| 7.8   | North  | ...                               |

And you want to train a model on data like that, but then report metrics like "RMSE predicting `sales` for offices in the `West` region"?

A reproducible example (working code that we could copy and run) showing exactly what you're doing would also increase the likelihood that you get a helpful answer here.

**Comment by Recepguel** [comment]
How about computing multiple feval's where each feval uses same metric but applies to a different output class?

**Comment by simpsus** [comment]
That sounds very interesting. I will give this a try. Thank you very much.

**Comment by simpsus** [comment]
@Recepguel I have trouble constructing an validation Dataset with multiple targets. 

In the docs (https://lightgbm.readthedocs.io/en/v3.3.5/pythonapi/lightgbm.Dataset.html#lightgbm.Dataset) it says label is one dimensional parameter. 

Do you have a hint if I should try to attach the additional targets to the dataset manually, like

```
ds = lgb.Dataset(data=features, label=training_target)
ds.eval_targets = pd.DataFrame(additional_targets)
```

Or should I try to include them in the features of the original set?

**Comment by jrvmalik** [comment]
@simpsus Sure add any attributes to that object you can do like
```
from __future__ import annotations

import numpy as np
import lightgbm as lgb


class Dataset(lgb.Dataset):

    def __init__(self, add_targets: np.ndarray, **kwargs) -> None:
        super(Dataset, self).__init__(**kwargs)
        self.add_targets = add_targets


data = Dataset(data=np.random.randn(30, 3),
               add_targets=np.random.randn(30, 2),
               init_score=np.zeros(30),
               label=np.random.randn(30))


def feval1(predictions: np.ndarray, dataset: Dataset) -> tuple[str, float, bool]:
    return 'metric1', np.mean(predictions - dataset.add_targets[:, 0]).item(), False


def feval2(predictions: np.ndarray, dataset: Dataset) -> tuple[str, float, bool]:
    return 'metric2', np.sum(predictions - dataset.add_targets[:, 1]).item(), False


lgb.train({'min_data_in_leaf': 1}, data, feval=[feval1, feval2], valid_sets=[data], valid_names=['train'])

```
